File: challenges/challenge-03/C03.ts
export function resolve() {
    const inputFile = Deno.readTextFileSync('colors.txt');
    const colors: string[] = JSON.parse(inputFile.replace(/'/g, '\"'))

    const zebras: string[][] = [
        [colors[0], colors[1]]
    ];

    let max_zebra: string[] = [];
    
    for (let i = 2; i < colors.length; i++) {
        const zebra = zebras[zebras.length - 1];
        const lastZebraColor = zebra[zebra.length - 1];

        if (lastZebraColor !== colors[i] && ( zebra.length === 1 || colors[i] === zebra[zebra.length - 2] ) ){
            zebra.push(colors[i])
            continue
        }

        if ( ( zebra.length === 1 && lastZebraColor === colors[i] ) || ( lastZebraColor !== colors[i] && colors[i] !== zebra[zebra.length - 2] ) ) {
            max_zebra = zebra.length > max_zebra.length ? [...zebra] : max_zebra;
            if (i+1 > colors.length - 1 || colors[i+1] !== lastZebraColor) {
                zebras.push([colors[i]])
                continue
            }

            zebras.push([lastZebraColor, colors[i]]);
            continue;
        }

        if (lastZebraColor === colors[i]) {
            max_zebra = zebra.length > max_zebra.length ? [...zebra] : max_zebra;
            zebras.push([ colors[i] ])
            continue
        }
        
    }

    return `${max_zebra.length}@${max_zebra.pop()}`
}

File: challenges/challenge-03/C03.py
from typing import List
from functools import reduce
import json

def resolve():

    input_file = open('colors.txt')
    input_file_str = reduce(lambda acc, line: acc + line.replace('\'', '"'), input_file, '')
    colors: List[str] = json.loads(input_file_str)

    colors_enumerate = enumerate(colors, 2)

    zebras = [ 
        [colors[0], colors[1]]
    ]
    max_zebra = []

    for i, color in colors_enumerate:

        if zebras[-1][-1] != color and (len(zebras[-1]) == 1 or color == zebras[-1][-2] ):
            zebras[-1].append(color)
            continue

        if ( len(zebras[-1]) == 1 and zebras[-1][-1] == color ) or ( zebras[-1][-1] != color and color != zebras[-1][-2] ):
            max_zebra = zebras[-1] if len(zebras[-1]) > len(max_zebra) else max_zebra
            if i+1 > len(colors) - 1 or colors[i+1] != zebras[-1][-1]:
                zebras.append([color])
                continue

            zebras.append([zebras[-1][-1], color])
            continue

        if zebras[-1][-1] == color:
            max_zebra = zebras[-1] if len(zebras[-1]) > len(max_zebra) else max_zebra
            zebras.append([ color ])
            continue
    
    return f"{len(max_zebra)}@{max_zebra[-1]}"
